Raise ArgumentError on Parameter.new

PhilibertDugas opened the pull request:
### Summary

When we pass a non-hash argument to `ActionController::Parameters.new` the error message is kinda cryptic:

```sh
[1] pry(main)> ActionController::Parameters.new("test")
NoMethodError: undefined method `with_indifferent_access' for "test":String
Did you mean?  with_indefinite_article
```

I think it would be clearer if we raise an argument error explaining clearly what we are doing wrong with something like this:

```sh
irb(main):007:0> ActionController::Parameters.new('test')
ArgumentError: invalid parameters given, the parameters should be a hash
```

PhilibertDugas committed:
Raise ArgumentError on Parameter.new

When we pass a non-hash argument to `ActionController::Parameters.new`
the error message is kinda cryptic:

```sh
[1] pry(main)> ActionController::Parameters.new("test")
NoMethodError: undefined method `with_indifferent_access' for "test":String
Did you mean?  with_indefinite_article
```

I think it would be clearer if we raise an argument error explaining
clearly what we are doing wrong with something like this:

```sh
irb(main):007:0> ActionController::Parameters.new('test')
ArgumentError: invalid parameters given, the parameters should be a hash
```

rafaelfranca commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+*   Raise ArgumentError on Parameter.new
```
Raise \`ArgumentError\` when passing anything that doesn't respond to \`with_indifferent_access\` to \`ActionController::Parameter.new\`.

rafaelfranca commented on actionpack/CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+*   Raise ArgumentError on Parameter.new
+
+    When we pass a non-hash argument to `ActionController::Parameters.new`
+    it now raises an `ArgumentError`
```
missing `.` in the end.

rafaelfranca commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -231,6 +231,7 @@ class Parameters
     #   params.permitted?  # => true
     #   Person.new(params) # => #<Person id: nil, name: "Francesco">
     def initialize(parameters = {})
+      raise ArgumentError, "invalid parameters given, the parameters should be a hash" unless parameters.is_a?(Hash)
```
I think it is better to check for `parameters.respond_to?(:with_indifferent_access)`

PhilibertDugas committed:
Update CHANGELOG description and use respond_to?

sgrif commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -231,6 +231,9 @@ class Parameters
     #   params.permitted?  # => true
     #   Person.new(params) # => #<Person id: nil, name: "Francesco">
     def initialize(parameters = {})
+      unless parameters.respond_to?(:with_indifferent_access)
```
Should we just `rescue NoMethodError` instead? Code is a bit less clear, but `respond_to?` is not free and this is a luke-warm code path.

rafaelfranca replied:
Yeah. I think that is better.

PhilibertDugas replied:
I'm up for it 👍 , mind explaining how `respond_to?` is more costly than rescuing? Trying to understand it better 😅

sgrif replied:
In the case where an actual error occurs, rescuing is more expensive. (`rescue` is essentially free, `raise` is the expensive bit). `respond_to?` isn't the most costly thing in the world, but it's non-zero cost. Basically doing `rescue` means that if no error occurs, we are doing no additional work.

PhilibertDugas replied:
ah good point! thanks for the clarification

matthewd replied:
> `rescue` is essentially free

Err.. rescue is quite expensive

meinac replied:
I agree with @matthewd, it's quite expensive and also this implementation hides the errors can occur inside of the `with_indifferent_access` method. If we do something wrong inside of this method which raises `NoMethodError` then we say `ArgumentError` which creates much more confusion than what it does now.

meinac replied:
Btw just in case of error so @sgrif is right on in normal case but this is not important than the possible confusion this implementation can lead.

PhilibertDugas replied:
Hmm, I guess the complexity / cost of raising an `ArgumentError` is not worth the potential confusion of a `NoMethodError` raised since this is rarely used in a user code? 🤔

meinac replied:
yes but imagine that, in case of a `NoMethodError` error in the method `with_indifferent_access` you catch this error and raise `ArgumentError`. Someone can override `with_indifferent_access` method and do some nasty stuff which can lead this error and in this case raising `ArgumentError` is much more confusing because even people will not be able to track the stack trace. I think we need to stick with your first implementation and check the responsibility of object with `respond_to?`.

PhilibertDugas replied:
Ah I see, yes that might get confusing 🤔 are you not worried about the `respond_to?` implementation in terms of performance?

meinac replied:
yes, I'm worried about it so I would prefer keep the method as it is in master now. Since this class is heavily used by Rails itself I think it's ok.

PhilibertDugas committed:
Rescue NoMethodError instead of early checks

Most of the time, the codepath will not fail, so checking everything for
a `.respond_to?` is an extra cost. By rescuing `NoMethodError` we are
not costing anything for most of the cases, only when there's an issue

matthewd commented:
If the new exception is going to say it should be a hash, one wonders if we shouldn't just make that the condition too.

But either way, this seems quite defensive, for a class I wouldn't normally expect to see instantiated by user code at all. What makes this particularly worthy of such protection?

PhilibertDugas commented:
Bear with me here, I was helping the migration of a project from Rails 5.0 --> Rails 5.1 and found some code that looked like this:
```ruby
def initialize(param)
  @param = ActionController::Parameters.new(param)
end
```
The constructor parameter `param` was  sometime a `Hash`, other times `ActionController::Parameters` directly. I understand this is not the best code & I need to fix this on my side, but I just felt like `NoMethodError: undefined method 'with_indifferent_access'` was not the best error message 😅 

There might be a better approach?

sgrif commented:
r? @matthewd 

I'm fine with this as is, but you seem to have concerns so I'll let you make the call

PhilibertDugas commented:
After discussion, looks like the drawbacks of adding this `ArgumentError` are greater than the gains
